modulus/distributed/utils.py

```diff
@@ -218,3 +219,293 @@ def _gather(input_, dim_, group=None):
+    dim: int = 0,
+    dst: int = 0,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed gather_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert 0 <= dst < comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+
+    if comm_size == 1:
+        return tensor
+
+    tensor_shape = list(tensor.shape)
+    tensor_list = [None] * comm_size
+
+    if rank == dst:
+        for r in range(comm_size):
+            req_list = [None] * comm_size
+            if r == dst:
+                tensor_list[dst] = tensor
+            else:
+                tensor_shape[dim] = sizes[r]
+                tensor_list[r] = torch.empty(
+                    tensor_shape,
+                    device=tensor.device,
+                    dtype=tensor.dtype,
+                )
+
+                req_list[r] = dist.irecv(tensor_list[r], src=r, group=group)
+
+        for r in range(comm_size):
+            if r != dst:
+                req_list[r].wait()
+
+        output = torch.cat(tensor_list, dim=dim)
+
+    else:
+        req = dist.isend(tensor, dst=dst, group=group)
+        req.wait()
+
+        tensor_shape[dim] = sum(sizes)
+        output = torch.empty(
+            tensor_shape,
+            device=tensor.device,
+            dtype=tensor.dtype,
+        )
+
+    return output
+
+
+def scatter_v_wrapper(
+    tensor: torch.Tensor,
+    sizes: List[int],
+    dim: int = 0,
+    src: int = 0,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed scatter_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert 0 <= src < comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+    local_size = sizes[rank]
+
+    assert global_size == tensor.size(dim)
+
+    tensor_shape = list(tensor.shape)
+    tensor_shape[dim] = sizes[rank]
+    output = torch.empty(
+        tensor_shape,
+        dtype=tensor.dtype,
+        device=tensor.device,
+    )
+
+    # dist.scatter doesn't support tensors of different shape
+    # so this implementation is using explicit send/recv combinations
+    scatter_list = None
+    if rank == src:
+        scatter_list = torch.split(tensor, sizes, dim=dim)
+        req_list = [None] * comm_size
+        for r in range(comm_size):
+            tensor_to_scatter_to_r = scatter_list[r]
+            if r == src:
+                output = tensor_to_scatter_to_r
+            else:
+                req_list[r] = dist.isend(tensor_to_scatter_to_r, dst=r, group=group)
+
+        for r in range(comm_size):
+            if r != src:
+                req_list[r].wait()
+
+    else:
+        req = dist.irecv(output, src=src, group=group)
+        req.wait()
+
+    return output
+
+
+def indexed_all_gather_wrapper(
```

Comment:
Why is this called `indexed_all_gather`? Isn't this like a neighbor all-to-all kinda like `MPI_Neighbor_Alltoallv`?

Reply:
Never was really sure how to call this to be honest. However, I agree that following MPI in that regard more closely could make sense, I'll change it to `all_to_all_v` then.

Reply:
Actually, I still think that indexed_all_gather or something similar would be more appropriate. 

What this routine basically does is something like
```
node 0 has data elements 0, ..., 9
node 1 has data elements 10, ..., 15
etc...

node 0 wants to gather 0, 5 from 0
node 1 wants to gather 1, 4, 5 from 0
node 2 wants to gather 6, 7, 9 from 0
etc... and similarly each node gathers certain values from all the remaining ranks
```

So technically, this is not just each rank gathering a variable amount of (contiguous) data elements from each other rank but a variable amount of arbitrary (non-contiguous) elements. Does this make sense?

Reply:
Yes, this makes sense. But I still think all_gather is somewhat confusing here. Typically, all_gather means at the end of the call, all ranks have the same output data which is not the case here right? I agree that there should be an indexed qualifier though because it is non-contiguous elements like you said. An `indexed_all_to_all_v` perhaps?

Reply:
That sounds better, I agree on the difference of `all_to_all_v` and `all_gather` here. 👍🏽